syllabus/curricula/plr/task_sampler.py

```diff
@@ -285,16 +363,76 @@ def sample(self, strategy=None):
+                        task = self._evaluate_unseen_level()
+                        episode_data = self.evaluate_task(task, self.eval_envs, self.action_value_fn, self.gamma,
+                                                          self.gae_lambda)
+                        self.update_with_episode_data(episode_data, self._average_gae)  # Update task scores
+
+                        # Check if we need to sample another unseen level
+                        num_unseen = (self.unseen_task_weights > 0).sum()
+                        proportion_seen = (self.num_tasks - num_unseen) / self.num_tasks
+                        if proportion_seen < self.rho or np.random.rand() >= proportion_seen:
+                            break
+
+                    return task
+                else:
+                    return self._sample_unseen_level()
+
         else:
-            raise NotImplementedError(f"Unsupported replay schedule: {self.replay_schedule}. Must be 'fixed' or 'proportionate'.")
+            raise NotImplementedError(
+                f"Unsupported replay schedule: {self.replay_schedule}. Must be 'fixed' or 'proportionate'.")
+
+    def update_with_episode_data(self, episode_data, score_function):
+        tasks = episode_data['tasks']
+        done = ~(episode_data['masks'] > 0)
+        total_steps, num_actors = episode_data['tasks'].shape[:2]
+
+        for actor_index in range(num_actors):
+            done_steps = done[:, actor_index].nonzero()[:total_steps, 0]
+            start_t = 0
+
+            for t in done_steps:
+                if not start_t < total_steps:
+                    break
+
+                if (t == 0):  # if t is 0, then this done step caused a full update of previous last cycle
+                    continue
+
+                task_idx_t = tasks[start_t, actor_index].item()
+
+                score_function_kwargs = {}
+                score_function_kwargs["returns"] = episode_data['returns'][start_t:t, actor_index]
+                score_function_kwargs["value_preds"] = episode_data['value_preds'][start_t:t, actor_index]
+                score = score_function(**score_function_kwargs)
+                num_steps = len(episode_data['tasks'][start_t:t, actor_index])
+                self.update_task_score(actor_index, task_idx_t, score, num_steps)
+
+                start_t = t.item()
+            if start_t < total_steps:
+                task_idx_t = tasks[start_t, actor_index].item()
+
+                score_function_kwargs = {}
+                score_function_kwargs["returns"] = episode_data['returns'][start_t:, actor_index]
+                score_function_kwargs["value_preds"] = episode_data['value_preds'][start_t:, actor_index]
+                score = score_function(**score_function_kwargs)
+                self._last_score = score
+                num_steps = len(episode_data['tasks'][start_t:, actor_index])
+                self._partial_update_task_score(actor_index, task_idx_t, score, num_steps)
```

Comment:
Could you reduce the code duplication between update_with_rollouts and this? Maybe create an inner function that does most of the computation, and a helper function that converts rollouts to an episode_data dictionary.

So update_with_rollouts will first move data into the episode_data dictionary and then call update_with_episode_data